tests/lib/rules/no-invalid-names.js

```diff
@@ -0,0 +1,232 @@
+                errors: [{
+                    messageId: `${objType}NameInvalidType`,
+                    data: {
+                        type: "ArrayExpression",
+                        name: "[]"
+                    }
+                }]
+            },
+            {
+                code: `${callee}({}, function () {});`,
+                output: null,
+                errors: [{
+                    messageId: `${objType}NameInvalidType`,
+                    data: {
+                        type: "ObjectExpression",
+                        name: "{}"
+                    }
+                }]
+            },
+            {
+                code: `${callee}(this, function () {});`,
+                output: null,
+                errors: [{
+                    messageId: `${objType}NameInvalidType`,
+                    data: {
+                        type: "ThisExpression",
+                        name: "this"
+                    }
+                }]
+            },
+            {
+                code: `${callee}(typeof foo, function () {});`,
+                output: null,
+                errors: [{
+                    messageId: `${objType}NameInvalidType`,
+                    data: {
+                        type: "UnaryExpression",
+                        name: "typeof foo"
+                    }
+                }]
+            },
+            {
+                code: `${callee}(void foo, function () {});`,
+                output: null,
+                errors: [{
+                    messageId: `${objType}NameInvalidType`,
+                    data: {
+                        type: "UnaryExpression",
+                        name: "void foo"
+                    }
+                }]
+            },
+            {
+                code: `${callee}(++foo, function () {});`,
+                output: null,
+                errors: [{
+                    messageId: `${objType}NameInvalidType`,
+                    data: {
+                        type: "UpdateExpression",
+                        name: "++foo"
+                    }
+                }]
+            },
+            {
+                code: `${callee}(foo + bar, function () {});`,
+                output: null,
+                errors: [{
+                    messageId: `${objType}NameInvalidType`,
+                    data: {
+                        type: "BinaryExpression",
+                        name: "foo + bar"
+                    }
+                }]
+            },
+            {
+                code: `${callee}(foo = "name", function () {});`,
+                output: null,
+                errors: [{
+                    messageId: `${objType}NameInvalidType`,
+                    data: {
+                        type: "AssignmentExpression",
+                        name: "foo = \"name\""
+                    }
+                }]
+            },
+            {
+                code: `${callee}(foo || "name", function () {});`,
```

Comment:
If somebody wants to decide on a string with `foo || "name"`, that should be allowed. It's a good way to set a fallback. I believe this should be a valid test case.

Reply:
I think we need to be careful to allow the user to dynamically generate the name anyway they would like. That's one thing I learned working on this plugin and on eslint-plugin-eslint-plugin in particular -- users can be very creative in how they generate their data, and unnecessarily restricting what types of statements they can use will end up causing headaches (and lots of bug reports) later.

Reply:
That's one approach, I suppose, but I'm also a fan of telling people to use disable statements when they want to work around a rule in a convoluted way.